### PR TITLE
Add const probing to delineated path type check

### LINKED ISSUE #3063: Assignment with call path does not work with constants
> I tested out with a script depending on a library with a pub const.
> 
> lib:
> 
> ```rust
> library test_lib;
> 
> pub const NUMBER: u64 = 10;
> ```
> 
> script that depends on the lib:
> 
> ```rust
> script;
> 
> fn main() {
>   let x = test_lib::NUMBER;
> }
> ```
> 
> and trying to build the script above, fails with:
> 
> ``` ^^^^^^ Could not find symbol "NUMBER" in this scope. ```
> 
> _Originally posted by @kayagokalp in https://github.com/FuelLabs/sway/pull/3016#discussion_r998063381_
>

### PR DESCRIPTION
**Pull request by kayagokalp**
closes #3063.

This enables usage of call paths that leads to const inside a declaration which would enable us to simplify given doc examples at #3016 . Example:

lib:

```rust
library test_lib;

pub const NUMBER: u64 = 10;
```

script that depends on the lib:

```rust
script;

fn main() {
  let x = test_lib::NUMBER;
}
```

is now possible.

I added const probing to the existing checks we have for `type_check_delineated_path` and tried to follow other checks there but my approach could be off too 😄

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/multiple_impl/json_abi_oracle.json [hunk 1]
```diff
@@ -20,7 +20,7 @@
       }
     },
     {
-      "logId": 165711,
```
Comment (kayagokalp) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/multiple_impl/json_abi_oracle.json:23]: Not sure the reason of this change 🤔

Comment (sezna) [reply test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/multiple_impl/json_abi_oracle.json:23]: @mohammadfawaz is working on stabilizing it. Small changes to the code result in all log ids being updated. Not ideal but hopefully temporary.

Comment (mohammadfawaz) [reply test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/multiple_impl/json_abi_oracle.json:23]: I knew this was gonna cause problems but I didn't know that every other PR is going to be affected 😢

**Review by mitchmindtree — COMMENTED** [review]
Looking great! Just needs a test or two :nerd_face:

**Comment by kayagokalp** [comment]
@mitchmindtree @Centril thanks for the review! Should be good to go now!

**Comment by mohammadfawaz** [comment]
Merge with `master` and resolved conflicts by reverting to the JSON ABIs on `master` which are now more stable.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,8 @@
+script;
+
+dep test_lib;
+
+fn main() -> u64 {
+  let x = test_lib::NUMBER;
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw:6]: Would be nice to also see an example with a longer path. You can try a constant from the standard library for example: `std::constants::ZERO_B256`. Also a constant with a non-primitive type such as `std::constants::BASE_ASSET_ID`. And then use `assert`s to check each of them if you want.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw [hunk 3]
```diff
@@ -0,0 +1,8 @@
+script;
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw:1]: The test should probably be formatted.

**Review by mohammadfawaz — COMMENTED** [review]
Just some minor nits.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw (hunk 2 above)
Comment (kayagokalp) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw]: I added the extra checks to the same test, would you prefer having them in a second separate test?

**Review by mohammadfawaz — APPROVED** [review]

File: test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw (hunk 2 above)
Comment (mohammadfawaz) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/const_decl_with_call_path/src/main.sw]: One test is perfectly fine :) Thanks!

**Review by mitchmindtree — APPROVED** [review]

**Pull request merged**